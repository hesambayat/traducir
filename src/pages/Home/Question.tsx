import { Shuffle } from '../../componentes'

interface QuestionProps {
  answers: number
  origin: string
  transforming: boolean
}

const Question = ({ answers, origin, transforming = false }: QuestionProps) => {
  return (
    <h1 className="home__question">
      {answers > 1 ? `What are ${answers} Spanish words for` : `What's Spanish for`} <Shuffle text={origin} live={!!transforming} />
    </h1>
  )
}

export default Question
