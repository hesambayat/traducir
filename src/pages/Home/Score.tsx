import { useModal } from '../../context/modal'

interface ScoreProps {
  total: number
  remaining: number
}

const Score = ({ total, remaining }: ScoreProps) => {
  const { setModal } = useModal()

  return (
    <header>
    <button className="home__score" onClick={setModal}>
      <span>?</span> You got {total - remaining}/{total}
    </button>
  </header>
  )
}

export default Score
