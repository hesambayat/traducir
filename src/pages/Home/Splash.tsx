import { useEffect, useState } from 'react'

const Splash = () => {
  const [className, setClassName] = useState('splash')

  useEffect(() => {
    const emit = setTimeout(() => {
      setClassName('splash splash--out')
    }, 1200)

    return () => clearTimeout(emit)
  }, [setClassName])

  return (
    <div className={className} />
  )
}

export default Splash
