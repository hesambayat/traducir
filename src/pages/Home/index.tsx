import { useCallback, useEffect, useReducer } from 'react'
import { Button, Keypad } from '../../componentes'
import { Congrats } from '../../pages'
import { useTranslationSource } from '../../hooks'
import Score from './Score'
import Question from './Question'
import words from '../../words.json'

enum Action {
  ADD = 'ADD',
  REMOVE = 'REMOVE'
}

const reducer = (state: number[][], action: { type: Action, payload?: number[][] }) => {
  switch (action.type) {
    case Action.ADD:
      return action.payload || []
    case Action.REMOVE:
      return []
    default:
      return state
  }
}

export const Home = () => {
  const [source, action, { remaining, total, status }] = useTranslationSource(words)
  const [highlights, dispatch] = useReducer(reducer, [])
  const onDone = useCallback((selected: number[][]) => {
    const { map } = source
    const payload = [...highlights, ...selected].filter((n, i, s) => s.indexOf(n) === i)
    const visited = payload.map(n => n.toString())
    const solved = map.flat().every(n => visited.includes(n.toString()))
    if (solved) {
      dispatch({ type: Action.REMOVE })
      action.shift()
    } else {
      dispatch({ type: Action.ADD, payload })
    }
  }, [source, highlights, action, dispatch])

  useEffect(() => {
    dispatch({ type: Action.REMOVE })
  }, [source, dispatch])

  if (!source) {
    return <Congrats />
  }

  return (
    <div className="home">
      <Score total={total} remaining={remaining} />
      <main>
        <Question answers={source.map.length} origin={source.origin} transforming={status !== 'idle'} />
        <Keypad {...source} highlights={highlights} onDone={onDone} transforming={status} />
        <div className="home__actions">
          <Button label="Skip &#10230;" disabled={remaining < 2 || status !== 'idle'} onClick={action.push} />
        </div>
      </main>
    </div>
  )
}

export default Home
