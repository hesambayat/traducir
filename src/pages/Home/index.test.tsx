import { render } from '@testing-library/react'
import Home from '.'

describe('renders home page', () => {
  const { asFragment } = render(<Home />)
  const dom = asFragment()

  test('should match the snapshot', () => {
    expect(dom).toMatchSnapshot()
  })
})
