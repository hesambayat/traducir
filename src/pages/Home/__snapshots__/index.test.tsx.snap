// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders home page should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="home"
  >
    <header>
      <button
        class="home__score"
      >
        <span>
          ?
        </span>
         You got 0/9
      </button>
    </header>
    <main>
      <h1
        class="home__question"
      >
        Find Spanish word for 
        <span>
          man
        </span>
      </h1>
      <div
        class="keypad"
        style="--grid: 8;"
      >
        <button
          class="key"
          data-testid="key"
        >
          i
        </button>
        <button
          class="key"
          data-testid="key"
        >
          q
        </button>
        <button
          class="key"
          data-testid="key"
        >
          í
        </button>
        <button
          class="key"
          data-testid="key"
        >
          l
        </button>
        <button
          class="key"
          data-testid="key"
        >
          n
        </button>
        <button
          class="key"
          data-testid="key"
        >
          n
        </button>
        <button
          class="key"
          data-testid="key"
        >
          m
        </button>
        <button
          class="key"
          data-testid="key"
        >
          ó
        </button>
        <button
          class="key"
          data-testid="key"
        >
          f
        </button>
        <button
          class="key"
          data-testid="key"
        >
          t
        </button>
        <button
          class="key"
          data-testid="key"
        >
          v
        </button>
        <button
          class="key"
          data-testid="key"
        >
          ñ
        </button>
        <button
          class="key"
          data-testid="key"
        >
          b
        </button>
        <button
          class="key"
          data-testid="key"
        >
          m
        </button>
        <button
          class="key"
          data-testid="key"
        >
          h
        </button>
        <button
          class="key"
          data-testid="key"
        >
          a
        </button>
        <button
          class="key"
          data-testid="key"
        >
          h
        </button>
        <button
          class="key"
          data-testid="key"
        >
          j
        </button>
        <button
          class="key"
          data-testid="key"
        >
          é
        </button>
        <button
          class="key"
          data-testid="key"
        >
          t
        </button>
        <button
          class="key"
          data-testid="key"
        >
          e
        </button>
        <button
          class="key"
          data-testid="key"
        >
          t
        </button>
        <button
          class="key"
          data-testid="key"
        >
          o
        </button>
        <button
          class="key"
          data-testid="key"
        >
          z
        </button>
        <button
          class="key"
          data-testid="key"
        >
          x
        </button>
        <button
          class="key"
          data-testid="key"
        >
          á
        </button>
        <button
          class="key"
          data-testid="key"
        >
          o
        </button>
        <button
          class="key"
          data-testid="key"
        >
          i
        </button>
        <button
          class="key"
          data-testid="key"
        >
          e
        </button>
        <button
          class="key"
          data-testid="key"
        >
          ñ
        </button>
        <button
          class="key"
          data-testid="key"
        >
          m
        </button>
        <button
          class="key"
          data-testid="key"
        >
          é
        </button>
        <button
          class="key"
          data-testid="key"
        >
          q
        </button>
        <button
          class="key"
          data-testid="key"
        >
          é
        </button>
        <button
          class="key"
          data-testid="key"
        >
          i
        </button>
        <button
          class="key"
          data-testid="key"
        >
          ó
        </button>
        <button
          class="key"
          data-testid="key"
        >
          q
        </button>
        <button
          class="key"
          data-testid="key"
        >
          s
        </button>
        <button
          class="key"
          data-testid="key"
        >
          b
        </button>
        <button
          class="key"
          data-testid="key"
        >
          s
        </button>
        <button
          class="key"
          data-testid="key"
        >
          c
        </button>
        <button
          class="key"
          data-testid="key"
        >
          u
        </button>
        <button
          class="key"
          data-testid="key"
        >
          m
        </button>
        <button
          class="key"
          data-testid="key"
        >
          y
        </button>
        <button
          class="key"
          data-testid="key"
        >
          v
        </button>
        <button
          class="key"
          data-testid="key"
        >
          l
        </button>
        <button
          class="key"
          data-testid="key"
        >
          r
        </button>
        <button
          class="key"
          data-testid="key"
        >
          x
        </button>
        <button
          class="key"
          data-testid="key"
        >
          ü
        </button>
        <button
          class="key"
          data-testid="key"
        >
          í
        </button>
        <button
          class="key"
          data-testid="key"
        >
          ó
        </button>
        <button
          class="key"
          data-testid="key"
        >
          m
        </button>
        <button
          class="key"
          data-testid="key"
        >
          o
        </button>
        <button
          class="key"
          data-testid="key"
        >
          t
        </button>
        <button
          class="key"
          data-testid="key"
        >
          e
        </button>
        <button
          class="key"
          data-testid="key"
        >
          k
        </button>
        <button
          class="key"
          data-testid="key"
        >
          a
        </button>
        <button
          class="key"
          data-testid="key"
        >
          g
        </button>
        <button
          class="key"
          data-testid="key"
        >
          r
        </button>
        <button
          class="key"
          data-testid="key"
        >
          n
        </button>
        <button
          class="key"
          data-testid="key"
        >
          n
        </button>
        <button
          class="key"
          data-testid="key"
        >
          ó
        </button>
        <button
          class="key"
          data-testid="key"
        >
          s
        </button>
        <button
          class="key"
          data-testid="key"
        >
          m
        </button>
      </div>
      <div
        class="home__actions"
      >
        <button
          class="button"
          data-testid="button"
        >
          Skip ⟶
        </button>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
